ndcube/ndcollection.py

```diff
@@ -0,0 +1,261 @@
+import collections.abc
+import copy
+import textwrap
+
+import numpy as np
+
+from ndcube import NDCube, NDCubeSequence
+from ndcube.utils.cube import convert_extra_coords_dict_to_input_format
+import ndcube.utils.collection as collection_utils
+
+__all__ = ["NDCollection"]
+
+
+class NDCollection(dict):
+    def __init__(self, key_data_pairs, aligned_axes=None, meta=None, **kwargs):
+        """
+        A class for holding and manipulating a collection of aligned NDCube or NDCubeSequences.
+
+        Parameters
+        ----------
+        data: sequence of `tuple`s of (`str`, `~ndcube.NDCube` or `~ndcube.NDCubeSequence`)
+            The names and data cubes/sequences to held in the collection.
+
+        aligned_axes: `tuple` of `int`, `tuple` of `tuple`s of `int`, 'all', or None, optional
+            Axes of each cube/sequence that are aligned in numpy order.
+            If elements are int, then the same axis numbers in all cubes/sequences are aligned.
+            If elements are tuples of ints, then must be one tuple for every cube/sequence.
+            Each element of each tuple gives the axes of each cube/sequence that are aligned.
+            If 'all', all axes are aligned in natural order, i.e. the 0th axes of all cubes
+            are aligned, as are the 1st, and so on.
+            Default=None
+
+        meta: `dict`, optional
+            General metadata for the overall collection.
+
+        Example
+        -------
+        Say the collection holds two NDCubes, each of 3 dimensions.
+        ``aligned_axes = (1, 2)``
+        means that axis 1 (0-based counting) of cube0 is aligned with axis 1 of cube1,
+        and axis 2 of cube0 is aligned with axis 2 of cube1.
+        However, if
+        ``aligned_axes = ((0, 1), (2, 1))``
+        then the first tuple corresponds to cube0 and the second with cube1.
+        This is interpretted as axis 0 of cube0 is aligned with axis 2 of cube1 while
+        axis 1 of cube0 is aligned with axis 1 of cube1.
+
+        """
+        # Enter data and metadata into object.
+        super().__init__(key_data_pairs)
+        self.meta = meta
+
+        # Convert aligned axes to required format.
+        if aligned_axes is not None:
+            keys, data = zip(*key_data_pairs)
+            # Sanitize aligned axes unless hidden kwarg indicates not to.
+            if kwargs.get("sanitize_inputs", True):
```

Comment:
We should probably validate we don't have any other kwargs other than this in `kwargs` as they don't go anywhere.